app/services/katello/candlepin/profiles_uploader.rb

```diff
@@ -0,0 +1,61 @@
+module Katello
+  module Candlepin
+    class ProfilesUploader
+      def self.upload(profile_string:, host: nil, host_id: nil)
+        profiles = JSON.parse(profile_string)
+        #free the huge string from the memory
+        profile_string = 'TRIMMED'.freeze
+        if host.nil?
+          Rails.logger.warn("Host with ID %s not found; continuing" % host_id)
+        elsif host.content_facet.nil? || host.content_facet.uuid.nil?
+          Rails.logger.warn("Host with ID %s has no content facet; continuing" % host_id)
+        elsif profiles.try(:has_key?, "deb_package_profile")
+          # remove this when deb_package_profile API is removed
+          payload = profiles.dig("deb_package_profile", "deb_packages") || []
+          import_deb_package_profile(host, payload)
+        else
+          module_streams = []
+          profiles.each do |profile|
+            payload = profile["profile"]
+            case profile["content_type"]
+            when "rpm"
+              PackageProfileUploader.import_package_profile_for_host(host_id, payload)
+            when "deb"
+              import_deb_package_profile(host, payload)
+            when "enabled_repos"
+              host.import_enabled_repositories(payload)
+            else
+              module_streams << payload
+            end
+          end
+
+          module_streams.each do |module_stream_payload|
+            import_module_streams(module_stream_payload, host)
+          end
+
+          ::Katello::Host::ContentFacet.trigger_applicability_generation(host.id)
```

Comment:
I think applicability regeneration should be decoupled from this class

Reply:
Okay, but that means I have to un-DRY it a bit and call it both from the controller and from the task.  Updated.  (or did you mean something else?)

Reply:
That's just what I meant! I see upload profiles and regen applicability as two steps in a workflow. Honestly it's not the end of the world to have it the way you did (the way it *was*) but I do like the explicit-ness of calling them both in serial